linalg/kernels.hpp

```diff
@@ -65,155 +68,183 @@ double Norml2(const int size, const T *data)
 template<typename TA, typename TB, typename TC>
 MFEM_HOST_DEVICE inline
-void Mult(const int ah, const int aw, const int bw,
-          const TB *B, const TC *C, TA *A)
+void Mult(const int Aheight, const int Awidth, const int Bwidth,
+          const TB *Bdata, const TC *Cdata, TA *Adata)
 {
-   const int ah_x_aw = ah*aw;
-   for (int i = 0; i < ah_x_aw; i++) { A[i] = 0.0; }
-   for (int j = 0; j < aw; j++)
+   const int ah_x_aw = Aheight * Awidth;
+   for (int i = 0; i < ah_x_aw; i++) { Adata[i] = 0.0; }
+   for (int j = 0; j < Awidth; j++)
    {
-      for (int k = 0; k < bw; k++)
+      for (int k = 0; k < Bwidth; k++)
       {
-         for (int i = 0; i < ah; i++)
+         for (int i = 0; i < Aheight; i++)
          {
-            A[i+j*ah] += B[i+k*ah] * C[k+j*bw];
+            Adata[i+j*Aheight] += Bdata[i+k*Aheight] * Cdata[k+j*Bwidth];
          }
       }
    }
 }
 
-// *****************************************************************************
-/** @brief Matrix vector multiplication: y = A x, where the matrix A is of size
-    @a height x @a width with data given by @a data. */
-template<typename TDATA, typename TX, typename TY>
+/** @brief Multiply a matrix of size @a Aheight x @a Awidth and data @a Adata
+    with the transpose of a matrix of size @a Bheight x @a Awidth and data @a
+    Bdata: A * Bt. Return the result in a matrix with data @a ABtdata. */
+template<typename TA, typename TB, typename TC>
 MFEM_HOST_DEVICE inline
-void MultV(const int height, const int width,
-           TDATA *data, const TX *x, TY *y)
+void MultABt(const int Aheight, const int Awidth, const int Bheight,
+             const TA *Adata, const TB *Bdata, TC *ABtdata)
 {
-   if (width == 0)
+   const int ah_x_bh = Aheight * Bheight;
+   for (int i = 0; i < ah_x_bh; i++) { ABtdata[i] = 0.0; }
+   for (int k = 0; k < Awidth; k++)
    {
-      for (int row = 0; row < height; row++)
+      TC *c = ABtdata;
+      for (int j = 0; j < Bheight; j++)
       {
-         y[row] = 0.0;
+         const double bjk = Bdata[j];
+         for (int i = 0; i < Aheight; i++)
+         {
+            c[i] += Adata[i] * bjk;
+         }
+         c += Aheight;
       }
-      return;
-   }
-   TDATA *d_col = data;
-   TX x_col = x[0];
-   for (int row = 0; row < height; row++)
-   {
-      y[row] = x_col*d_col[row];
+      Adata += Aheight;
+      Bdata += Bheight;
    }
-   d_col += height;
-   for (int col = 1; col < width; col++)
+}
+
+/// Compute the spectrum of the matrix of size dim with given @a data, returning
+/// the eigenvalues in the array @a lambda and the eigenvectors in the array @a
+/// vec (listed consecutively).
+template<int dim>
+void CalcEigenvalues(const double *data, double *lambda, double *vec);
+
+/// Return the i'th singular value of the matrix of size dim with given @a data.
+template<int dim>
+double CalcSingularvalue(const double *data, const int i);
+
+
+// Utility functions for CalcEigenvalues and CalcSingularvalue
+namespace internal
```

Comment:
New namespace `mfem::kernels::internal` for the utility functions

Reply:
Are you sure about the comma in `EXCLUDE_SYMBOLS`?
The documentation script now fails in an `internal` class of the `mem_manager`.

Reply:
Let me take a look...

Reply:
You were right, fixed